net/bcp38/files/run.sh

```diff
@@ -22,83 +26,91 @@ add_bcp38_rule()
 	local action="$2"
 
 	if [ "$action" == "nomatch" ]; then
-		ipset add "$IPSET_NAME" "$subnet" nomatch
+		nft add element "$FAMILY" "$TABLE" "$NOMATCHSET" { "$subnet" }
 	else
-		ipset add "$IPSET_NAME" "$subnet"
+		nft add element "$FAMILY" "$TABLE" "$MATCHSET" { "$subnet" }
 	fi
 }
 
-detect_upstream()
+detect_upstream_subnet()
 {
 	local interface="$1"
 
 	subnets=$(ip route show dev "$interface"  | grep 'scope link' | awk '{print $1}')
 	for subnet in $subnets; do
-		# ipset test doesn't work for subnets, so strip out the subnet part
-		# and test for that; add as exception if there's a match
-		addr=$(echo $subnet | sed 's|/[0-9]\+$||')
-		ipset test "$IPSET_NAME" $addr 2>/dev/null && add_bcp38_rule $subnet nomatch
+		#test for that; add as exception if there's a match
+		nft get element "$FAMILY" "$TABLE" "$MATCHSET" { $subnet } >/dev/null 2>/dev/null && add_bcp38_rule $subnet nomatch
 	done
 }
 
 run() {
-    	local section="$1"
-    	local enabled
+	local section="$1"
+	local enabled
 	local interface
+	local priority
 	local detect_upstream
 	config_get_bool enabled "$section" enabled 0
 	config_get interface "$section" interface
 	config_get detect_upstream "$section" detect_upstream
+	config_get priority "$section" priority "2"
```

Comment:
Does this really need to be configurable? Presumably we know the priority of the default openwrt firewall chains, and we always want to hook in before this? So I think just setting the priority as a fixed value at the top of the file (with the chain name) is fine...

Reply:
Actually the priority 2 adds the bcp38 chains **after** the standard fw4 chains. 
I believe, this is a thing that can't be decided as it depends on the users situation. With nftables we place the hook either in front of or after the fw4 chains.
I the old fw3 implementation we needed to use the user_rules which have been placed before the rest of the fw3 rules but have been part of the fw4 rules.
With nftables chains are processed one after each other and only if a packet is accepted it is passed to the next chain. If it passes all chains accepted, it is accepted. Any drop or reject skips processing of further chains.
So the question is: do we need/want to filter this before or after the fw4 firewall? Placing it before means, each packet (that is not offloaded) is passing through the bcp filter before it is processed by fw4 firewall. Placing it after means, only packets that have not been already dropped in fw4 will be checked. In either way the filtering is applied correctly. 
So it depends on how many of those packets come in and I believe that fw4 is filtering more packets than bcp38, so I prefer to place this filter after fw4. But with the configuration option, the user may choose another solution that fits better for him. 
 
Having dedicated chains means that the filter does nor interfere with fw4 in any way as fw4 firewall only acts on fw4 table.

Reply:
Right, if the semantics is such that a packet needs an accept verdict by all chains, it's fine to execute the BCP filter after the regular firewall. I still don't think there's any real need to make it configurable, though; we can always add a knob if someone turns out to have a real use case for that...